### PR TITLE
[Speculative decoding] Improve n-gram efficiency

### PR DESCRIPTION
**Pull request by comaniac**
This PR refactors the ngram (PLD) implementation to make it more efficient. Specifically, in ngram worker
1. Do not use `.all()` when n-gram size is 1.
2. Keep all tensors on GPU all the time and avoid using Python native lists.
3. When a sequence has no match, return None instead of padding 0's.
4. When the speculative token indices exceed the input length, simply `clamp` them instead of padding 0's.
5. Change `ngram_prompt_lookup_min` to be inclusive, so the minimum valid value becomes `1`. IMO this is more intuitive to users, as `--ngram-prompt-lookup-min=0` looks like we are disabling this.

Also in top1_proposer, this PR adds a feature to allow draft worker to not propose speculative tokens for certain sequences. Before this PR, unmatched sequence still proposes all 0 tokens and they will be scored by the target model as well.

Test cases are fixed accordingly to reflect the changes. The new added features are covered already.
So far I haven't observed an obvious speedup with these changes except for ngram size 1, but these changes should be nice to have. Please let me know if you have any opinion.

cc @cadedaniel @leiwen83 @LiuXiaoxuanPKU 

**BEFORE SUBMITTING, PLEASE READ THE CHECKLIST BELOW AND FILL IN THE DESCRIPTION ABOVE**

---

<details>

<summary><b> PR Checklist (Click to Expand) </b></summary>

<p>Thank you for your contribution to vLLM! Before submitting the pull request, please ensure the PR meets the following criteria. This helps vLLM maintain the code quality and improve the efficiency of the review process.</p>

<h3>PR Title and Classification</h3>
<p>Only specific types of PRs will be reviewed. The PR title is prefixed appropriately to indicate the type of change. Please use one of the following:</p>
<ul>
    <li><code>[Bugfix]</code> for bug fixes.</li>
    <li><code>[CI/Build]</code> for build or continuous integration improvements.</li>
    <li><code>[Doc]</code> for documentation fixes and improvements.</li>
    <li><code>[Model]</code> for adding a new model or improving an existing model. Model name should appear in the title.</li>
    <li><code>[Frontend]</code> For changes on the vLLM frontend (e.g., OpenAI API server, <code>LLM</code> class, etc.) </li>
    <li><code>[Kernel]</code> for changes affecting CUDA kernels or other compute kernels.</li>
    <li><code>[Core]</code> for changes in the core vLLM logic (e.g., <code>LLMEngine</code>, <code>AsyncLLMEngine</code>, <code>Scheduler</code>, etc.)</li>
    <li><code>[Hardware][Vendor]</code> for hardware-specific changes. Vendor name should appear in the prefix (e.g., <code>[Hardware][AMD]</code>).</li>
    <li><code>[Misc]</code> for PRs that do not fit the above categories. Please use this sparingly.</li>
</ul>
<p><strong>Note:</strong> If the PR spans more than one category, please include all relevant prefixes.</p>

<h3>Code Quality</h3>

<p>The PR need to meet the following code quality standards:</p>

<ul>
    <li>We adhere to <a href="https://google.github.io/styleguide/pyguide.html">Google Python style guide</a> and <a href="https://google.github.io/styleguide/cppguide.html">Google C++ style guide</a>.</li>
    <li>Pass all linter checks. Please use <a href="https://github.com/vllm-project/vllm/blob/main/format.sh"><code>format.sh</code></a> to format your code.</li>
    <li>The code need to be well-documented to ensure future contributors can easily understand the code.</li>
    <li>Include sufficient tests to ensure the project to stay correct and robust. This includes both unit tests and integration tests.</li>
    <li>Please add documentation to <code>docs/source/</code> if the PR modifies the user-facing behaviors of vLLM. It helps vLLM user understand and utilize the new features or changes.</li>
</ul>

<h3>Notes for Large Changes</h3>
<p>Please keep the changes as concise as possible. For major architectural changes (>500 LOC excluding kernel/data/config/test), we would expect a GitHub issue (RFC) discussing the technical design and justification. Otherwise, we will tag it with <code>rfc-required</code> and might not go through the PR.</p>

<h3>What to Expect for the Reviews</h3>

<p>The goal of the vLLM team is to be a <i>transparent reviewing machine</i>. We would like to make the review process transparent and efficient and make sure no contributor feel confused or frustrated. However, the vLLM team is small, so we need to prioritize some PRs over others. Here is what you can expect from the review process: </p>

<ul>
    <li> After the PR is submitted, the PR will be assigned to a reviewer. Every reviewer will pick up the PRs based on their expertise and availability.</li>
    <li> After the PR is assigned, the reviewer will provide status update every 2-3 days. If the PR is not reviewed within 7 days, please feel free to ping the reviewer or the vLLM team.</li>
    <li> After the review, the reviewer will put an <code> action-required</code> label on the PR if there are changes required. The contributor should address the comments and ping the reviewer to re-review the PR.</li>
    <li> Please respond to all comments within a reasonable time frame. If a comment isn't clear or you disagree with a suggestion, feel free to ask for clarification or discuss the suggestion.
 </li>
</ul>

<h3>Thank You</h3>

<p> Finally, thank you for taking the time to read these guidelines and for your interest in contributing to vLLM. Your contributions make vLLM a great tool for everyone! </p>

</details>

### CODE REVIEW

File: vllm/spec_decode/top1_proposer.py
Comment (cadedaniel) [inline vllm/spec_decode/top1_proposer.py:88]: * If we skip a sequence in this case, does it still do normal decoding?
* Can we add a test that hits this path? lazily, we could do an ngram e2e test over sufficiently long generation. more precisely, we could do a unit test on top1 proposer or ngram worker.

**Review by cadedaniel — COMMENTED** [review]
LGTM, nice improvement. some question

File: vllm/spec_decode/ngram_worker.py
Comment (cadedaniel) [inline vllm/spec_decode/ngram_worker.py:108]: this line looks sus, from a performance perspective

File: vllm/spec_decode/top1_proposer.py
Comment (comaniac) [reply vllm/spec_decode/top1_proposer.py:88]: * It should just do the same thing as other sequences with zero proposal lengths (due to max context length or auto-disabling). Plz correct if I'm wrong.
* An existing test already covers this case: https://github.com/vllm-project/vllm/pull/4724/files#diff-4ce22933dc7ebf3c65c3b299481cfafe4e3493a37dd32c5bf4df54d0dc70d18bL66. The first sequence was having `proposals.proposal_lens[0] = proposal_len` and `proposals.proposal_token_ids[0] = [0,0,...]`, and now having `proposals.proposal_lens[0]=0` and `proposals.proposal_token_ids[0] = [-1,-1,...]`.

File: vllm/spec_decode/ngram_worker.py
Comment (comaniac) [reply vllm/spec_decode/ngram_worker.py:108]: Good point. It might transfer the value back to CPU. I'll double check.

File: vllm/spec_decode/top1_proposer.py
Comment (cadedaniel) [inline vllm/spec_decode/top1_proposer.py:80]: Can we move the contents of this branch to own function for readability and separation?

File: vllm/spec_decode/top1_proposer.py
Comment (cadedaniel) [inline vllm/spec_decode/top1_proposer.py:82]: We should assert that `maybe_sampler_output` is of shape [batch] and not [proposal_len]

File: vllm/spec_decode/ngram_worker.py
Comment (comaniac) [reply vllm/spec_decode/ngram_worker.py]: I've checked and yes it does have to transfer the value to CPU for the branch. However, I'm not sure if the following alternative would be better:

If we want to avoid GPU-CPU transfer at all during this entire process, we cannot have if-conditions in this loop. It means instead of

```python
# Now: Approach 1
matches = (windows[:-1] == ngram_tensor).all(dim=-1)
first_match = matches.max(dim=-1)
if first_match.values.item(): # GPU -> CPU (tensor with shape (1,))
    # generate speculative token ids and logprobs based on first_match.indices
    break
```

we have to do

```python
# Alternative: Approach 2
matches = (windows[:-1] == ngram_tensor).all(dim=-1)
first_match = matches.max(dim=-1)
# generate speculative token ids and logprobs based on first_match.indices
# even the "first match" isn't really a match (first_match.value == False) 
```

Approach 2 avoids CPU-GPU communication, but introduces additional overheads when there's no ngram match, because we will generate proposals and logprobs for tokens IDs `input_ids[:k]` anyways. This is more like random guess, so the performance highly depends on the acceptance rate. On the other hand, if we adopt this approach, we don't need to change Top1Proposer because every sequence is guaranteed to have a proposal.

I don't have strong feeling about which one works better in general tho. WDYT?

File: vllm/spec_decode/top1_proposer.py
Comment (cadedaniel) [inline vllm/spec_decode/top1_proposer.py:88]: This is O(bs^2), suggest using `set` to make it O(bs)

File: vllm/spec_decode/ngram_worker.py
Comment (Yard1) [reply vllm/spec_decode/ngram_worker.py]: nonzero forces CPU<->GPU sync as well - https://pytorch.org/docs/stable/generated/torch.nonzero.html#torch-nonzero

I'd suggest instead doing something like `num_matches = (temp != 0).sum()`?

Comment (Yard1) [reply vllm/spec_decode/ngram_worker.py]: but I guess you need to sync here anyway... would probably need a custom kernel to avoid that.

Comment (comaniac) [reply vllm/spec_decode/ngram_worker.py]: Ah I see...but `(temp != 0).sum()` results in something like `num_matches = tensor(1, device='cuda:0')`, and we still need `num_matches.tolist()` to make that comparison.

Then I guess still, the only way I could think of to avoid CPU-GPU sync is removing the if-condition, meaning that we don't check if ngram really matches or not. If there's no match, `first_match = matches.max(dim=-1)` will point to the first input token. Then the final performance highly depends on how lucky we are.

Comment (comaniac) [reply vllm/spec_decode/ngram_worker.py]: Hmm...actually, this would work poorly either, because we will always make a proposal with the maximum ngram whatever it matches or not due to no conditional break...

Comment (Yard1) [reply vllm/spec_decode/ngram_worker.py]: yeah I think the only way is to have the condition on CUDA itself, so a custom kernel.

Comment (Yard1) [reply vllm/spec_decode/ngram_worker.py]: > Hmm...actually, this would work poorly either, because we will always make a proposal with the maximum ngram whatever it matches or not due to no conditional break...

Honestly might not be an issue? Those GPU ops here should be quite fast. Ideally we'd delay the sync as long as possible - I think that would still give benefits compared to having the sync in.

File: vllm/spec_decode/top1_proposer.py
Comment (cadedaniel) [reply vllm/spec_decode/top1_proposer.py:88]: > It should just do the same thing as other sequences with zero proposal lengths (due to max context length or auto-disabling). Plz correct if I'm wrong.

I follow now, sounds good!

> An existing test already covers this case:

Great!

Comment (cadedaniel) [reply vllm/spec_decode/top1_proposer.py:88]: (there's an argument here that for small BS, set may take just as much time. not sure).

File: vllm/spec_decode/ngram_worker.py
Comment (cadedaniel) [reply vllm/spec_decode/ngram_worker.py]: the spec decode framework has a few such synchronizations. we can fix this one here but until they're all solved the benefit won't be that large. (if you want to pioneer a solution e.g. lower conditional logic to kernel then fine by me 😄  )

**Review by cadedaniel — COMMENTED** [review]
small comments

File: vllm/spec_decode/ngram_worker.py
Comment (cadedaniel) [reply vllm/spec_decode/ngram_worker.py]: ways to minimize the cost of a synchronization is to batch them. e.g. do a synchronization once for all iterations of the for loop. not sure if that's possible in your ngram matching logic.

Comment (Yard1) [reply vllm/spec_decode/ngram_worker.py]: yes, batching is what I thought of would work here

Comment (comaniac) [reply vllm/spec_decode/ngram_worker.py]: There are two challenges for batching:
1. The length of each sequence differs.
2. The first match ngram size of each sequence differs.

Point 1 means we have to pad input IDs if we batch sequences for ngram matching. Point 2 means we have to perform ngram matching from max ngram size to min ngram size for all sequences (i.e., no break). These 2 points also increase memory footprint, because we keep 1) all input IDs with padding (but I guess this tensor should not be too large?), and 2) all slicing windows (generated by `.unfold`).

I'll think a bit more about this today and get you back, but we may unblock this PR if there's no obvious solution shortly.

Comment (cadedaniel) [reply vllm/spec_decode/ngram_worker.py]: Yeah don’t mean to block the PR on this, we can optimize further later

File: vllm/spec_decode/top1_proposer.py
Comment (comaniac) [reply vllm/spec_decode/top1_proposer.py:88]: It doesn't hurt to use set tho. Will do.

**Review by cadedaniel — APPROVED** [review]
approving preemptively (am OOO next week)

**Review by Yard1 — APPROVED** [review]

**Pull request merged**
